core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java

```diff
@@ -162,133 +122,139 @@ public int getTopologyId() {
+    * @param entries a {@code Flowable} with all the entries that need to be sent
+    */
+   public CompletionStage<Void> execute(Flowable<InternalCacheEntry<Object, Object>> entries) {
+      CompletableFuture<Void> taskFuture = new CompletableFuture<>();
       try {
-         // TODO: need to change to SDC.forEachSegment
-         // send data container entries
-         for (InternalCacheEntry ice : dataContainer) {
-            Object key = ice.getKey();  //todo [anistor] should we check for expired entries?
-            int segmentId = keyPartitioner.getSegment(key);
-            if (segments.contains(segmentId) && !ice.isL1Entry()) {
-               InternalCacheEntry entry = mapEntryFromDataContainer.apply(ice, entryFactory);
-               if (entry != null) {
-                  sendEntry(entry, segmentId);
-               }
-            }
-         }
-
-         AdvancedCacheLoader<Object, Object> stProvider = persistenceManager.getStateTransferProvider();
-         if (stProvider != null) {
-            try {
-               Flowable.fromPublisher(stProvider.entryPublisher(k -> !dataContainer.containsKey(k), true, true))
-                     .blockingForEach(me -> {
-                        int segmentId = keyPartitioner.getSegment(me.getKey());
-                        if (segments.contains(segmentId)) {
-                           try {
-                              InternalCacheEntry entry = mapEntryFromStore.apply(me, entryFactory);
-                              if (entry != null) {
-                                 sendEntry(entry, segmentId);
-                              }
-                           } catch (CacheException e) {
-                              log.failedLoadingValueFromCacheStore(me.getKey(), e);
-                           }
-                        }
-                     });
-            } catch (CacheException e) {
-               log.failedLoadingKeysFromCacheStore(e);
-            }
-         }
-
-         // send the last chunk of all segments
-         sendEntries(true);
+         AtomicReference<List<InternalCacheEntry<Object, Object>>> batchRef =
+            new AtomicReference<>(Collections.emptyList());
+         entries.buffer(chunkSize)
+                .takeUntil(batch -> cancelled)
+                .concatMapCompletable(batch -> {
+                   // Send the previous batch, not the current one
+                   // This allows us to mark all the segments as finished in the same RPC with the
+                   // last batch
+                   List<InternalCacheEntry<Object, Object>> previousBatch = batchRef.getAndSet(batch);
+                   if (previousBatch.isEmpty())
+                      return Completable.complete();
+
+                   return completionStageToCompletable().apply(sendEntries(previousBatch, false));
+                }, 1)
+                .subscribe(new CompletableObserver() {
```

Comment:
I would say we could use http://reactivex.io/RxJava/javadoc/io/reactivex/Completable.html#subscribe-io.reactivex.functions.Action-io.reactivex.functions.Consumer- instead now, but I won't let that block integration.

Reply:
I was actually using that at first, but I switched to the anonymous class to cut back on the lambdas :)

TBH maintaining the subscription was easier that way, because that overload is returning a `Disposable`, but I didn't want to mess with subscriptions any more.